README.md
# Hockeyapp android Deploy

The new Hockeyapp android Deploy step.

Can be run directly with the [bitrise CLI](https://github.com/bitrise-io/bitrise),
just `git clone` this repository, `cd` into it's folder in your Terminal/Command Line
and call `bitrise run test`.

*Check the `bitrise.yml` file for required inputs which have to be
added to your `.bitrise.secrets.yml` file!*

## Run the tests in Docker, with `docker-compose`

You can call `docker-compose run --rm app bitrise run test` to run the test
inside the Bitrise Android Docker image.


CHANGELOG.md
## Changelog (Current version: 1.2.1)

-----------------

### 1.2.1 (2017 Dec 07)

* [a209447] Bump version to 1.2.1
* [c869f2d] Add description that there can be multiple paths (#12)

### 1.2.0 (2017 Dec 04)

* [da08649] Bump version to 1.2.0
* [312e23b] Support deploying multiple APKs (#11)

### 1.1.4 (2017 Oct 12)

* [c923b8c] Prepare for 1.1.4

### 1.1.3 (2017 Oct 10)

* [9a1e044] Prepare for 1.1.3
* [dca09ae] added project type tags

### 1.1.2 (2017 Oct 09)

* [f95d1b3] Prepare for 1.1.2
* [9c10ebf] Fix typo (#8)
* [a487be2] RELEASE_VERSION: 1.1.1

### 1.1.0 (2016 Sep 15)

* [92a861a] error printing fixes
* [51eac39] error check fix (#5)
* [c1ab33c] Merge pull request #3 from godrei/multiline_notes
* [cef1cdb] step.go, multiline notes
* [f55b23d] step.yml : "package ID" instead of "bundle ID"

### 0.9.0 (2015 Dec 16)

* [997ec12] osx added as available host_os_tags
* [a76615b] bitrise.yml : marked the step to "always run" (run_if: true) for quick local testing
* [99a43e8] step.yml revision : step summary & description update ; host_os_tags update ; run_if: IsCI ; and removed the unnecessary opts from outputs
* [f3dee1b] step.yml : title fix
* [91d041c] step.sh : when no file found at the provided apk_path print the path (debug)
* [5a4bf18] docker-compose.yml and README for quick testing in docker (with docker-compose)
* [6d72fea] updated bitrise.yml for quick testing
* [65acdee] bitrise.yml : share-this-step workflow
* [d24f824] step.yml revision : HockeyApp App ID description ; apk_path is set to `$BITRISE_APK_PATH` by default
* [aa6a465] Merge pull request #1 from godrei/master
* [55f1be4] log fix
* [32657f6] step

### 1.0.1 (2015 Dec 08)


-----------------

Updated: 2017 Dec 07

main_test.go
package main

import (
	"testing"
	"github.com/stretchr/testify/require"
	"os"
	"github.com/bitrise-io/go-utils/command"
	"io/ioutil"
	"github.com/bitrise-io/go-utils/log"
)

func TestValidateConfigApkPathDontUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.ApkPath = "/dev/null"
	require.NoError(t, configs.validate())
}

func TestValidateConfigApkPathUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.IsUseApkPathList = "true"
	configs.ApkPath = "/dev/null"
	require.Error(t, configs.validate())
}

func TestValidateConfigApkPathListUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.IsUseApkPathList = "true"
	configs.ApkPathList = []string{"/dev/null"}
	require.NoError(t, configs.validate())
}

func TestValidateConfigApkPathListDontUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.ApkPathList = []string{"/dev/null"}
	require.Error(t, configs.validate())
}

func TestValidateConfigApkPathAndListListUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.IsUseApkPathList = "true"
	configs.ApkPath = "/dev/null"
	configs.ApkPathList = []string{"/dev/null"}
	require.NoError(t, configs.validate())
}

func TestActualApkPathListUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.IsUseApkPathList = "true"
	configs.ApkPath = "single apk"
	configs.ApkPathList = []string{"apk from the list"}
	require.Equal(t, configs.ApkPathList, configs.getActualApkPathList())
}

func TestActualApkPathListDontUseList(t *testing.T) {
	configs := createDummyConfigs()
	configs.ApkPath = "single apk"
	configs.ApkPathList = []string{"apk from the list"}
	require.Equal(t, []string{configs.ApkPath}, configs.getActualApkPathList())
}

func TestExportSingleValue(t *testing.T) {
	envstore, err := setupTempEnvstore()
	require.NoError(t, err)
	defer func() {
		err := os.Remove(envstore)
		if err != nil {
			log.Errorf("Could not delete temporary file %s, error %s", envstore, err)
		}
	}()

	require.NoError(t, os.Setenv("ENVMAN_ENVSTORE_PATH", envstore))
	exportSingleEnvironmentOrLogWarn("SAMPLE_KEY", "sample value")

	environment, err := command.New("envman", "print").RunAndReturnTrimmedOutput()
	require.NoError(t, err)

	require.Contains(t, environment, "SAMPLE_KEY")
	require.Contains(t, environment, "sample value")
}

func TestExportMultipleValues(t *testing.T) {
	envstore, err := setupTempEnvstore()
	require.NoError(t, err)
	defer func() {
		err := os.Remove(envstore)
		if err != nil {
			log.Warnf("Could not delete temporary file %s, error %s", envstore, err)
		}
	}()

	require.NoError(t, os.Setenv("ENVMAN_ENVSTORE_PATH", envstore))
	exportEnvironmentSliceOrLogWarn("SAMPLE_KEY", []string{"sample value", "sample value 2"})

	environment, err := command.New("envman", "print").RunAndReturnTrimmedOutput()
	require.NoError(t, err)

	require.Contains(t, environment, "SAMPLE_KEY")
	require.Contains(t, environment, "sample value|sample value 2")
}

func createDummyConfigs() ConfigsModel {
	return ConfigsModel{
		IsUseApkPathList:"false",
		APIToken:"token",
		NotesType:"0",
		Notify:"2",
		Status:"2",
		Mandatory:"false",
		ApkPath:"",
		MappingPath:"",
		ApkPathList: []string{""},
	}
}

func setupTempEnvstore() (string, error) {
	file, err := ioutil.TempFile("", "envstore")
	if err != nil {
		return "", err
	}
	return file.Name(), nil
}


step.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"os"
	"os/exec"
	"strings"
)

// -----------------------
// --- Constants
// -----------------------

const (
	hockeyAppDeployStatusKey     = "HOCKEYAPP_DEPLOY_STATUS"
	hockeyAppDeployStatusSuccess = "success"
	hockeyAppDeployStatusFailed  = "failed"
	hockeyAppDeployPublicURLKey  = "HOCKEYAPP_DEPLOY_PUBLIC_URL"
	hockeyAppDeployBuildURLKey   = "HOCKEYAPP_DEPLOY_BUILD_URL"
	hockeyAppDeployConfigURLKey  = "HOCKEYAPP_DEPLOY_CONFIG_URL"
)

// -----------------------
// --- Models
// -----------------------

// ResponseModel ...
type ResponseModel struct {
	ConfigURL string `json:"config_url"`
	PublicURL string `json:"public_url"`
	BuildURL  string `json:"build_url"`
}

// -----------------------
// --- Functions
// -----------------------

func logFail(format string, v ...interface{}) {
	if err := exportEnvironmentWithEnvman(hockeyAppDeployStatusKey, hockeyAppDeployStatusFailed); err != nil {
		logWarn("Failed to export %s, error: %s", hockeyAppDeployStatusKey, err)
	}

	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[31;1m%s\x1b[0m\n", errorMsg)
	os.Exit(1)
}

func logWarn(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[33;1m%s\x1b[0m\n", errorMsg)
}

func logInfo(format string, v ...interface{}) {
	fmt.Println()
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("\x1b[34;1m%s\x1b[0m\n", errorMsg)
}

func logDetails(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("  %s\n", errorMsg)
}

func logDone(format string, v ...interface{}) {
	errorMsg := fmt.Sprintf(format, v...)
	fmt.Printf("  \x1b[32;1m%s\x1b[0m\n", errorMsg)
}

func genericIsPathExists(pth string) (os.FileInfo, bool, error) {
	if pth == "" {
		return nil, false, errors.New("No path provided")
	}
	fileInf, err := os.Stat(pth)
	if err == nil {
		return fileInf, true, nil
	}
	if os.IsNotExist(err) {
		return nil, false, nil
	}
	return fileInf, false, err
}

// IsPathExists ...
func IsPathExists(pth string) (bool, error) {
	_, isExists, err := genericIsPathExists(pth)
	return isExists, err
}

func exportEnvironmentWithEnvman(keyStr, valueStr string) error {
	envman := exec.Command("envman", "add", "--key", keyStr)
	envman.Stdin = strings.NewReader(valueStr)
	envman.Stdout = os.Stdout
	envman.Stderr = os.Stderr
	return envman.Run()
}

func createRequest(url string, fields, files map[string]string) (*http.Request, error) {
	var b bytes.Buffer
	w := multipart.NewWriter(&b)

	// Add fields
	for key, value := range fields {
		if err := w.WriteField(key, value); err != nil {
			return nil, err
		}
	}

	// Add files
	for key, file := range files {
		f, err := os.Open(file)
		if err != nil {
			return nil, err
		}
		fw, err := w.CreateFormFile(key, file)
		if err != nil {
			return nil, err
		}
		if _, err = io.Copy(fw, f); err != nil {
			return nil, err
		}
	}

	w.Close()

	req, err := http.NewRequest("POST", url, &b)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", w.FormDataContentType())

	return req, nil
}

// -----------------------
// --- Main
// -----------------------

func main() {
	//
	// Validate options
	apkPath := os.Getenv("apk_path")
	mappingPath := os.Getenv("mapping_path")
	apiToken := os.Getenv("api_token")
	appID := os.Getenv("app_id")
	notes := os.Getenv("notes")
	notesType := os.Getenv("notes_type")
	notify := os.Getenv("notify")
	status := os.Getenv("status")
	tags := os.Getenv("tags")
	commitSHA := os.Getenv("commit_sha")
	buildServerURL := os.Getenv("build_server_url")
	repositoryURL := os.Getenv("repository_url")

	// mandatory handling, with backward compatibility
	//  0 - not mandatory (default)
	//  1 - mandatory
	mandatory := "0"
	if os.Getenv("mandatory") == "1" || os.Getenv("mandatory") == "true" {
		mandatory = "1"
	}

	logInfo("Configs:")
	logDetails("apk_path: %s", apkPath)
	logDetails("mapping_path: %s", mappingPath)
	logDetails("api_token: ***")
	logDetails("app_id: %s", appID)
	logDetails("notes: %s", notes)
	logDetails("notes_type: %s", notesType)
	logDetails("notify: %s", notify)
	logDetails("status: %s", status)
	logDetails("mandatory: %s", mandatory)
	logDetails("tags: %s", tags)
	logDetails("commit_sha: %s", commitSHA)
	logDetails("build_server_url: %s", buildServerURL)
	logDetails("repository_url: %s", repositoryURL)

	if apkPath == "" {
		logFail("Missing required input: apk_path")
	}
	if exist, err := IsPathExists(apkPath); err != nil {
		logFail("Failed to check if apk (%s) exist, error: %#v", apkPath, err)
	} else if !exist {
		logFail("No apk found to deploy. Specified path was: %s", apkPath)
	}

	if mappingPath != "" {
		if exist, err := IsPathExists(mappingPath); err != nil {
			logFail("Failed to check if mapping (%s) exist, error: %#v", mappingPath, err)
		} else if !exist {
			logFail("No mapping found to deploy. Specified path was: %s", mappingPath)
		}
	}

	if apiToken == "" {
		logFail("No App api_token provided as environment variable. Terminating...")
	}

	//
	// Create request
	logInfo("Performing request")

	requestURL := "https://rink.hockeyapp.net/api/2/apps/upload"
	if appID != "" {
		requestURL = fmt.Sprintf("https://rink.hockeyapp.net/api/2/apps/%s/app_versions/upload", appID)
	}

	fields := map[string]string{
		"notes":            notes,
		"notes_type":       notesType,
		"notify":           notify,
		"status":           status,
		"mandatory":        mandatory,
		"tags":             tags,
		"commit_sha":       commitSHA,
		"build_server_url": buildServerURL,
		"repository_url":   repositoryURL,
	}

	files := map[string]string{
		"ipa": apkPath,
	}
	if mappingPath != "" {
		files["dsym"] = mappingPath
	}

	request, err := createRequest(requestURL, fields, files)
	if err != nil {
		logFail("Failed to create request, error: %#v", err)
	}
	request.Header.Add("X-HockeyAppToken", apiToken)

	client := http.Client{}
	response, requestErr := client.Do(request)

	defer response.Body.Close()
	contents, readErr := ioutil.ReadAll(response.Body)

	//
	// Process response

	// Error
	if requestErr != nil {
		if readErr != nil {
			logWarn("Failed to read response body, error: %#v", readErr)
		} else {
			logInfo("Response:")
			logDetails("status code: %d", response.StatusCode)
			logDetails("body: %s", string(contents))
		}
		logFail("Performing request failed, error: %#v", requestErr)
	}

	if response.StatusCode < 200 || response.StatusCode > 300 {
		if readErr != nil {
			logWarn("Failed to read response body, error: %#v", readErr)
		} else {
			logInfo("Response:")
			logDetails("status code: %d", response.StatusCode)
			logDetails("body: %s", string(contents))
		}
		logFail("Performing request failed, status code: %d", response.StatusCode)
	}

	// Success
	logDone("Request succed")

	logInfo("Response:")
	logDetails("status code: %d", response.StatusCode)
	logDetails("body: %s", contents)

	if readErr != nil {
		logFail("Failed to read response body, error: %#v", readErr)
	}

	var responseModel ResponseModel
	if err := json.Unmarshal([]byte(contents), &responseModel); err != nil {
		logFail("Failed to parse response body, error: %#v", err)
	}

	fmt.Println()
	if responseModel.PublicURL != "" {
		logDone("Public URL: %s", responseModel.PublicURL)
	}
	if responseModel.BuildURL != "" {
		logDone("Build (direct download) URL: %s", responseModel.BuildURL)
	}
	if responseModel.ConfigURL != "" {
		logDone("Config URL: %s", responseModel.ConfigURL)
	}

	if err := exportEnvironmentWithEnvman(hockeyAppDeployStatusKey, hockeyAppDeployStatusSuccess); err != nil {
		logFail("Failed to export %s, error: %#v", hockeyAppDeployStatusKey, err)
	}

	if err := exportEnvironmentWithEnvman(hockeyAppDeployPublicURLKey, responseModel.PublicURL); err != nil {
		logFail("Failed to export %s, error: %#v", hockeyAppDeployPublicURLKey, err)
	}

	if err := exportEnvironmentWithEnvman(hockeyAppDeployBuildURLKey, responseModel.BuildURL); err != nil {
		logFail("Failed to export %s, error: %#v", hockeyAppDeployBuildURLKey, err)
	}

	if err := exportEnvironmentWithEnvman(hockeyAppDeployConfigURLKey, responseModel.ConfigURL); err != nil {
		logFail("Failed to export %s, error: %#v", hockeyAppDeployConfigURLKey, err)
	}
}


main.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"mime/multipart"
	"net/http"
	"os"
	"strings"

	"github.com/bitrise-io/depman/pathutil"
	"github.com/bitrise-io/go-utils/command"
	"github.com/bitrise-io/go-utils/log"
)

const (
	hockeyAppDeployStatusKey     = "HOCKEYAPP_DEPLOY_STATUS"
	hockeyAppDeployStatusSuccess = "success"
	hockeyAppDeployStatusFailed  = "failed"

	hockeyAppDeployPublicURLKey = "HOCKEYAPP_DEPLOY_PUBLIC_URL"
	hockeyAppDeployBuildURLKey  = "HOCKEYAPP_DEPLOY_BUILD_URL"
	hockeyAppDeployConfigURLKey = "HOCKEYAPP_DEPLOY_CONFIG_URL"

	hockeyAppDeployPublicURLKeyList = "HOCKEYAPP_DEPLOY_PUBLIC_URL_LIST"
	hockeyAppDeployBuildURLKeyList  = "HOCKEYAPP_DEPLOY_BUILD_URL_LIST"
	hockeyAppDeployConfigURLKeyList = "HOCKEYAPP_DEPLOY_CONFIG_URL_LIST"
)

var configs ConfigsModel

// ConfigsModel ...
type ConfigsModel struct {
	ApkPath        []string
	MappingPath    string
	APIToken       string
	AppID          string
	Notes          string
	NotesType      string
	Notify         string
	Status         string
	Tags           string
	CommitSHA      string
	BuildServerURL string
	RepositoryURL  string
	Mandatory      string
}

func createConfigsModelFromEnvs() ConfigsModel {

	mandatory := os.Getenv("mandatory")
	if mandatory == "1" || mandatory == "true" {
		mandatory = "1"
	} else {
		mandatory = "0"
	}

	apkPath := []string{}
	for _, pth := range strings.Split(os.Getenv("apk_path"), "|") {
		if pth != "" {
			apkPath = append(apkPath, pth)
		}
	}

	return ConfigsModel{
		ApkPath:        apkPath,
		MappingPath:    os.Getenv("mapping_path"),
		APIToken:       os.Getenv("api_token"),
		AppID:          os.Getenv("app_id"),
		Notes:          os.Getenv("notes"),
		NotesType:      os.Getenv("notes_type"),
		Notify:         os.Getenv("notify"),
		Status:         os.Getenv("status"),
		Tags:           os.Getenv("tags"),
		CommitSHA:      os.Getenv("commit_sha"),
		BuildServerURL: os.Getenv("build_server_url"),
		RepositoryURL:  os.Getenv("repository_url"),
		Mandatory:      mandatory,
	}
}

func (configs ConfigsModel) print() {
	fmt.Println()
	log.Infof("Configs:")
	log.Printf(" - ApkPath: %s", configs.ApkPath)
	log.Printf(" - MappingPath: %s", configs.MappingPath)
	log.Printf(" - APIToken: %s", configs.APIToken)
	log.Printf(" - AppID: %s", configs.AppID)
	log.Printf(" - Notes: %s", configs.Notes)
	log.Printf(" - NotesType: %s", configs.NotesType)
	log.Printf(" - Notify: %s", configs.Notify)
	log.Printf(" - Status: %s", configs.Status)
	log.Printf(" - Tags: %s", configs.Tags)
	log.Printf(" - CommitSHA: %s", configs.CommitSHA)
	log.Printf(" - BuildServerURL: %s", configs.BuildServerURL)
	log.Printf(" - RepositoryURL: %s", configs.RepositoryURL)
	log.Printf(" - Mandatory: %s", configs.Mandatory)
}

func (configs ConfigsModel) validate() error {
	if len(configs.ApkPath) == 0 {
		return errors.New("no ApkPath parameter specified")
	}

	for _, apkPath := range configs.ApkPath {
		if exist, err := pathutil.IsPathExists(apkPath); err != nil {
			return fmt.Errorf("failed to check if ApkPath exist at: %s, error: %v", apkPath, err)
		} else if !exist {
			return fmt.Errorf("apkPath not exist at: %s", apkPath)
		}
	}

	required := map[string]string{
		"APIToken":  configs.APIToken,
		"NotesType": configs.NotesType,
		"Notify":    configs.Notify,
		"Status":    configs.Status,
		"Mandatory": configs.Mandatory,
	}
	for k, v := range required {
		if v == "" {
			return fmt.Errorf("no %s parameter specified", k)
		}
	}

	if configs.MappingPath != "" {
		if exist, err := pathutil.IsPathExists(configs.MappingPath); err != nil {
			return fmt.Errorf("failed to check if MappingPath exist at: %s, error: %v", configs.MappingPath, err)
		} else if !exist {
			return fmt.Errorf("mappingPath not exist at: %s", configs.MappingPath)
		}
	}

	return nil
}

// ResponseModel ...
type ResponseModel struct {
	ConfigURL string `json:"config_url"`
	PublicURL string `json:"public_url"`
	BuildURL  string `json:"build_url"`
}

func exportEnvironmentWithEnvman(keyStr, valueStr string) error {
	cmd := command.New("envman", "add", "--key", keyStr)
	cmd.SetStdin(strings.NewReader(valueStr))
	return cmd.Run()
}

func createRequest(url string, fields, files map[string]string) (*http.Request, error) {
	var b bytes.Buffer
	w := multipart.NewWriter(&b)

	for key, value := range fields {
		if err := w.WriteField(key, value); err != nil {
			return nil, err
		}
	}

	for key, file := range files {
		f, err := os.Open(file)
		if err != nil {
			return nil, err
		}
		fw, err := w.CreateFormFile(key, file)
		if err != nil {
			return nil, err
		}
		if _, err = io.Copy(fw, f); err != nil {
			return nil, err
		}
	}

	if err := w.Close(); err != nil {
		return nil, err
	}

	req, err := http.NewRequest("POST", url, &b)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", w.FormDataContentType())

	return req, nil
}

func deploy(apkPath string) (ResponseModel, error) {
	fmt.Println()
	log.Infof("Performing request")

	requestURL := "https://rink.hockeyapp.net/api/2/apps/upload"
	if configs.AppID != "" {
		requestURL = fmt.Sprintf("https://rink.hockeyapp.net/api/2/apps/%s/app_versions/upload", configs.AppID)
	}

	fields := map[string]string{
		"notes":            configs.Notes,
		"notes_type":       configs.NotesType,
		"notify":           configs.Notify,
		"status":           configs.Status,
		"mandatory":        configs.Mandatory,
		"tags":             configs.Tags,
		"commit_sha":       configs.CommitSHA,
		"build_server_url": configs.BuildServerURL,
		"repository_url":   configs.RepositoryURL,
	}

	files := map[string]string{
		"ipa": apkPath,
	}
	if configs.MappingPath != "" {
		files["dsym"] = configs.MappingPath
	}

	request, err := createRequest(requestURL, fields, files)
	if err != nil {
		return ResponseModel{}, fmt.Errorf("Failed to create request, error: %v", err)
	}

	request.Header.Add("X-HockeyAppToken", configs.APIToken)
	client := http.Client{}
	response, err := client.Do(request)
	if err != nil {
		return ResponseModel{}, fmt.Errorf("Performing request failed, error: %v", err)
	}
	defer func() {
		if err := response.Body.Close(); err != nil {
			log.Warnf("Failed to close response body, error: %v", err)
		}
	}()

	contents, readErr := ioutil.ReadAll(response.Body)
	if readErr != nil {
		return ResponseModel{}, fmt.Errorf("Failed to read response body, error: %v", readErr)
	} else if response.StatusCode < 200 || response.StatusCode > 300 {
		return ResponseModel{}, fmt.Errorf("Performing request failed, status code: %d", response.StatusCode)
	}

	log.Donef("Request succeeded")
	fmt.Println()
	log.Infof("Response:")
	log.Printf(" status code: %d", response.StatusCode)
	log.Printf(" body: %s", contents)

	responseModel := ResponseModel{}
	if err := json.Unmarshal([]byte(contents), &responseModel); err != nil {
		return ResponseModel{}, fmt.Errorf("Failed to parse response body, error: %v", err)
	}
	return responseModel, nil
}

func contains(list []string, item string) bool {
	for _, i := range list {
		if i == item {
			return true
		}
	}
	return false
}

func main() {
	configs = createConfigsModelFromEnvs()
	configs.print()
	if err := configs.validate(); err != nil {
		log.Errorf("Issue with input: %s", err)
		os.Exit(1)
	}

	configURLs := []string{}
	buildURLs := []string{}
	publicURLs := []string{}

	for _, apkPath := range configs.ApkPath {
		responseModel, err := deploy(apkPath)
		if err != nil {
			log.Errorf("Hockeyapp deploy failed: %v", err)
			if err := exportEnvironmentWithEnvman(hockeyAppDeployStatusKey, hockeyAppDeployStatusFailed); err != nil {
				log.Warnf("Failed to export %s, error: %v", hockeyAppDeployStatusKey, err)
			}
			os.Exit(1)
		}
		if responseModel.ConfigURL != "" && !contains(configURLs, responseModel.ConfigURL) {
			configURLs = append(configURLs, responseModel.ConfigURL)
			log.Donef("Config URL: %s", responseModel.ConfigURL)
		}
		if responseModel.BuildURL != "" && !contains(buildURLs, responseModel.BuildURL) {
			buildURLs = append(buildURLs, responseModel.BuildURL)
			log.Donef("Build (direct download) URL: %s", responseModel.BuildURL)
		}
		if responseModel.PublicURL != "" && !contains(publicURLs, responseModel.PublicURL) {
			publicURLs = append(publicURLs, responseModel.PublicURL)
			log.Donef("Public URL: %s", responseModel.PublicURL)
		}
	}

	outputs := map[string]string{
		hockeyAppDeployStatusKey:        hockeyAppDeployStatusSuccess,
		hockeyAppDeployConfigURLKeyList: strings.Join(configURLs, "|"),
		hockeyAppDeployBuildURLKeyList:  strings.Join(buildURLs, "|"),
		hockeyAppDeployPublicURLKeyList: strings.Join(publicURLs, "|"),
	}
	if len(configURLs) > 0 {
		outputs[hockeyAppDeployConfigURLKey] = configURLs[len(configURLs)-1]
	}
	if len(buildURLs) > 0 {
		outputs[hockeyAppDeployBuildURLKey] = buildURLs[len(buildURLs)-1]
	}
	if len(publicURLs) > 0 {
		outputs[hockeyAppDeployPublicURLKey] = publicURLs[len(publicURLs)-1]
	}

	for k, v := range outputs {
		if err := exportEnvironmentWithEnvman(k, v); err != nil {
			log.Warnf("Failed to export %s, error: %v", k, err)
		}
	}
}
